src/application.jl

```diff
@@ -17,7 +17,11 @@ function ldiv!(F::ILUFactorization, y::AbstractVector)
     transposed_backward_substitution!(F.U, y)
 end
 
-(\)(F::ILUFactorization, y::AbstractVector) = ldiv!(F, copy(y))
+function (\)(F::ILUFactorization{Tv}, x::AbstractVector{Tv2}) where {Tv,Tv2}
+    Tv3 = promote_type(Tv,Tv2)
+    y = Tv3.(copy(x))
+    ldiv!(F, y)
```

Comment:
```suggestion
    Tv3 = typeof((zero(Tv)*zero(Tv2) + zero(Tv)*zero(Tv2))/one(Tv))
    y = similar(x, Tv3)
    copyto!(y, x)
    ldiv!(F, y)
```

something like this is done in Julia's LinearAlgebra. Maybe a better solution is to use Julia's UnitUpperTriangular / LowerTriangular types though, but I'm happy to get this fix in first.